compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -326,7 +326,15 @@ object PatternMatcher {
           sym.isAllOf(SyntheticCase) && sym.owner.is(Scala2x)
 
         if (isSyntheticScala2Unapply(unapp.symbol) && caseAccessors.length == args.length)
-          matchArgsPlan(caseAccessors.map(ref(scrutinee).select(_)), args, onSuccess)
+          def tupleSel(sym: Symbol) = ref(scrutinee).select(sym)
+          def tupleApp(i: Int) = // manually inlining the call to NonEmptyTuple#apply, because it's an inline method
+            ref(defn.RuntimeTuplesModule)
+              .select(defn.RuntimeTuples_apply)
+              .appliedTo(ref(scrutinee), Literal(Constant(i)))
+              .cast(args(i).tpe.widen)
+          val isGenericTuple = !defn.isTupleNType(tree.tpe)
```

Comment:
I had a look at this and the issue is that we are getting a `Int *: Int *: EmptyTuple` for which `isTupleNType` returns `false`.

```suggestion
          val isGenericTuple = defn.tupleArity(tree.tpe, Definitions.MaxTupleArity).isEmpty
```

```scala
  // Add to definitions
  def tupleArity(tp: Type, bound: Int = Int.MaxValue)(using Context): Option[Int] = {
    @tailrec def rec(tp: Type, arity: Int): Option[Int] = tp.normalized.dealias match
      case _ if arity > bound => None
      case tp: AppliedType if defn.PairClass == tp.classSymbol => rec(tp.args(1), arity + 1)
      case tp: AppliedType if defn.isTupleClass(tp.tycon.classSymbol) && tp.args.size <= bound => Some(tp.args.size)
      case tp: TermRef if tp.symbol == defn.EmptyTupleModule => Some(arity)
      case _ => None
    rec(tp.stripTypeVar, 0)
  }
```

Unfortunately more things fail after that change

Reply:
Thanks Nico.  What are you testing?  That the compiler bootstraps?

Not sure I'll be able to get to this this week, btw.

Reply:
I think the case for tuple class types should be this (include arity in case of `Int *: Tuple2[String, Boolean]`):
```scala
case tp: AppliedType if defn.isTupleClass(tp.tycon.classSymbol) && tp.args.size <= bound => Some(arity + tp.args.size) 
```